agent/engine/docker_image_manager.go

```diff
@@ -303,14 +309,147 @@ func (imageManager *dockerImageManager) removeUnusedImages(ctx context.Context)
+	nonECSContainersIDs, err := imageManager.nonECSContainersIDs(ctx)
+	if err != nil {
+		fmt.Errorf("Cannoe get nonECSContainersIDs Error")
+	}
+	var nonECSContainerRemoveAvailableIDs []string
+	for _, id := range nonECSContainersIDs {
+		response, _ := imageManager.client.InspectContainer(ctx, id, dockerclient.InspectContainerTimeout)
+		if response.State.Status == "exited" && time.Now().Sub(response.State.FinishedAt) > imageManager.nonECSContainerCleanupWaitDuration {
+			nonECSContainerRemoveAvailableIDs = append(nonECSContainerRemoveAvailableIDs, id)
+		}
+	}
+	var numNonECSContainerDeleted = 0
+	for _, id := range nonECSContainerRemoveAvailableIDs {
+		if numNonECSContainerDeleted == imageManager.numNonECSContainersToDelete {
+			break
+		}
+		err := imageManager.client.RemoveContainer(ctx, id, dockerclient.RemoveContainerTimeout)
+		if err == nil {
+			numNonECSContainerDeleted++
+		} else {
+			seelog.Info("nonECS Container cannot be removed %v", id)
+			continue
+		}
+	}
+}
+
+func (imageManager *dockerImageManager) nonECSContainersIDs(ctx context.Context) ([]string, error) {
+	ECSContainersIDs := imageManager.state.GetAllContainerIDs()
+	var allContainersIDs []string
+	listContainersResponse := imageManager.client.ListContainers(ctx, true, dockerclient.ListContainersTimeout)
+	if listContainersResponse.Error != nil {
+		return nil, fmt.Errorf("listContainersResponse Error")
+	}
+	for _, dockerID := range listContainersResponse.DockerIDs {
+		allContainersIDs = append(allContainersIDs, dockerID)
+	}
+	nonECSContainersIDs := exclude(allContainersIDs, ECSContainersIDs)
+	return nonECSContainersIDs, nil
+}
+
+type ImageWithSize struct {
+	ImageName string
+	Size      int64
+}
+
+func (imageManager *dockerImageManager) removeNonECSImages(ctx context.Context, nonECSImagesNumToDelete int) {
+	var nonECSImageNames = imageManager.nonECSImagesNames(ctx)
+	var nonECSImageNamesRemoveEligible []string
+	for _, nonECSImage := range nonECSImageNames {
+		if !isInExclusionList(nonECSImage, imageManager.imageCleanupExclusionList) {
+			nonECSImageNamesRemoveEligible = append(nonECSImageNamesRemoveEligible, nonECSImage)
+		}
+	}
+
+	var ImageWithSizeList []ImageWithSize
+	for _, imageName := range nonECSImageNamesRemoveEligible {
+		resp, _ := imageManager.client.InspectImage(imageName)
+		ImageWithSizeList = append(ImageWithSizeList, ImageWithSize{imageName, resp.Size})
+	}
+	// we want to sort images with size ascending
+	sort.Slice(ImageWithSizeList, func(i, j int) bool {
+		return ImageWithSizeList[i].Size < ImageWithSizeList[j].Size
+	})
+
+	// we will remove the remaining nonECSImages in each performPeriodicImageCleanup call()
+	var numImagesAlreadyDelete = 0
+	for _, kv := range ImageWithSizeList {
+		if numImagesAlreadyDelete == nonECSImagesNumToDelete {
+			break
+		}
+		err := imageManager.client.RemoveImage(ctx, kv.ImageName, dockerapi.RemoveImageTimeout)
+		if err != nil {
+			seelog.Info("nonECS image cannot be removed %v", kv.ImageName)
```

Comment:
this should be Error level and we probably should log the error like this: seelog.Errorf("Error removing Image %s - %v", kv.ImageName, err)